libp2p/signed_envelope.nim

```diff
@@ -116,3 +118,49 @@ proc getField*(pb: ProtoBuffer, field: int,
       ok(true)
     else:
       err(ProtoError.IncorrectBlob)
+
+type
+  SignedPayload*[T] = object
+    # T needs to have .encode(), .decode(), .payloadType(), .domain()
```

Comment:
Could this constraint be enforced by typing `T`, eg `T: Payload` such that

```nim
type Payload = object

method encode*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")

method decode*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")

method payloadType*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")

method domain*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")
```
and signed typed would need to be derived from `Payload`, eg
```nim
type PeerRecord = object of Payload
  peerId*: PeerId
  seqNo*: uint64
  addresses*: seq[AddressInfo]
```

Reply:
The goal was to keep PeerRecord (or whatever T) "generic", ie not bound to the Envelope

I tried using a concept, works pretty well, though the error messages when the type does not match are not clear at all

Reply:
concepts are on the no-no list: unsupported and scheduled to be redone completely

Reply:
Ok, removed, worst case scenario it won't compile

Reply:
Could we instead wrap `PeerRecord` in a `Payload` type to keep `PeerRecord` from being modified, and creating a type that can be used in a constrained `SignedPayload` generic?

```nim
type Payload = object

method encode*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")

method decode*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")

method payloadType*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")

method domain*(self: Payload) {.base.} =
  doAssert(false, "Not implemented")
```
```nim
type PeerRecordPayload = object of Payload
  record: PeerRecord
```
```nim
type
  SignedPayload*[T: Payload] = object
```

Reply:
Sure, but the goal of this system was to avoid boilerplate, so that would bring us back to square one.
Currently, if you try to use the SignedPayload on a non-compatible type, you will get errors like:
```nim
SignedPayload[int].decode(xxx)
```


```
testsigned_envelope.nim([line of the decode], 28) template/generic instantiation of `decode` from here
signed_envelope.nim(156, 43) Error: type mismatch: got <typedesc[int]>
but expected one of:
proc payloadDomain(T: typedesc[DummyPayload]): string
  first type mismatch at position: 1
  required type for T: typedesc[DummyPayload]
  but expression 'T' is of type: typedesc[int]
```

Which I think is not too bad

Worst case, we can add some static asserts to check that the type is compatible

But I don't think this is necessary: `SignedPayload` should not be used by end-users, instead they should use the types created by the library maintainer (eg, `SignedPeerRecord`), and the library maintainer should make sure that their type work in the CI or other